src/features/virtualDocumentTracker.ts

```diff
@@ -0,0 +1,97 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import {workspace, TextDocument} from 'vscode';
+import {OmniSharpServer} from '../omnisharp/server';
+import * as serverUtils from '../omnisharp/utils';
+import { FileChangeType } from '../omnisharp/protocol';
+import { IDisposable } from '../Disposable';
+import CompositeDisposable from '../CompositeDisposable';
+import { EventStream } from '../EventStream';
+import { DocumentSynchronizationFailure } from '../omnisharp/loggingEvents';
+
+function trackCurrentVirtualDocuments(server: OmniSharpServer, eventStream: EventStream) {
+    let registration = server.onProjectAdded(async () => {
+        for (let i = 0; i < workspace.textDocuments.length; i++) {
+            let document = workspace.textDocuments[i];
+    
+            if (!shouldIgnoreDocument(document, server)) {
+                await openVirtualDocument(document, server, eventStream);
+            }
+        }
+
+        registration.dispose();
```

Comment:
I'm not sure what the exact semantics of the NodeJS EventEmitter are. Is it possible for this handler to reenter? If so you have a race on disposing the registration/could process the virtual documents more than once, no?

Reply:
maybe the vscode semantics are different but in normal node / JavaScript it's single threaded thus making races impossible.

Reply:
Ah, actually, the await yields execution. I can see how things can get invoked more than once unintentionally.